File: opensource/kernel/src/main/java/com/sos/scheduler/engine/kernel/order/jobchain/Node.java
package com.sos.scheduler.engine.kernel.order.jobchain;

import com.sos.scheduler.engine.kernel.scheduler.AbstractHasPlatform;
import com.sos.scheduler.engine.kernel.scheduler.Platform;
import com.sos.scheduler.engine.kernel.cppproxy.NodeC;
import com.sos.scheduler.engine.kernel.cppproxy.NodeCI;
import com.sos.scheduler.engine.kernel.cppproxy.Order_queue_nodeC;
import com.sos.scheduler.engine.kernel.order.OrderState;
import com.sos.scheduler.engine.cplusplus.runtime.Sister;
import com.sos.scheduler.engine.cplusplus.runtime.SisterType;

/**
 *
 * @author Zschimmer.sos
 */
public class Node extends AbstractHasPlatform implements Sister {
    private final NodeCI nodeC;


    Node(Platform platform, NodeCI nodeC) {
        super(platform);
        this.nodeC = nodeC;
    }


    @Override public final void onCppProxyInvalidated() {}


    public final OrderState getOrderState() { return OrderState.of(nodeC.string_order_state()); }
    public final OrderState getNextState() { return OrderState.of(nodeC.string_next_state()); }
    public final OrderState getErrorState() { return OrderState.of(nodeC.string_error_state()); }


    public static class Type implements SisterType<Node, NodeC> {
        @Override public Node sister(NodeC proxy, Sister context) {
            NodeCI nodeCI = proxy;
            return nodeCI instanceof Order_queue_nodeC?
                Order_queue_nodeC.sisterType.sister((Order_queue_nodeC)nodeCI, context)
              : new Node(Platform.of(context), proxy);
        }
    }
}

File: opensource/kernel/src/main/scala/com/sos/scheduler/engine/kernel/order/jobchain/Node.scala
package com.sos.scheduler.engine.kernel.order.jobchain

import com.google.inject.Injector
import com.sos.scheduler.engine.cplusplus.runtime.Sister
import com.sos.scheduler.engine.cplusplus.runtime.annotation.ForCpp
import com.sos.scheduler.engine.data.jobchain.{NodeOverview, JobChainPath, JobChainNodeAction, JobChainNodePersistentState}
import com.sos.scheduler.engine.data.order.OrderState
import com.sos.scheduler.engine.kernel.cppproxy.NodeCI

/** @author Joacim Zschimmer */
@ForCpp
abstract class Node extends Sister {

  protected val cppProxy: NodeCI
  protected def injector: Injector

  def onCppProxyInvalidated() {}

  final def persistentState =
    new JobChainNodePersistentState(jobChainPath, orderState, action)

  def overview: NodeOverview

  final def jobChainPath =
    JobChainPath(cppProxy.job_chain_path)

  final def orderState =
    OrderState(cppProxy.string_order_state)

  final def nextState =
    OrderState(cppProxy.string_next_state)

  final def errorState =
    OrderState(cppProxy.string_error_state)

  final def action =
    JobChainNodeAction.ofCppName(cppProxy.string_action)

  final def action_=(o: JobChainNodeAction) {
    cppProxy.set_action_string(o.toCppName)
  }
}
